test_journal.py

```diff
@@ -0,0 +1,373 @@
+    testing.setUp(settings=settings)
+    req = testing.DummyRequest()
+
+    def cleanup():
+        testing.tearDown()
+
+    request.addfinalizer(cleanup)
+
+    return req
+
+
+"""
+    Below:
+
+    Tests that use our new 'auth_req' feature
+"""
+
+
+def test_do_login_success(auth_req):
+    from journal import do_login
+    auth_req.params = {'username': 'admin', 'password': 'secret'}
+    assert do_login(auth_req)
+
+
+def test_do_login_bad_pass(auth_req):
+    from journal import do_login
+    auth_req.params = {'username': 'admin', 'password': 'wrong'}
+    assert not do_login(auth_req)
+
+
+def test_do_login_bad_user(auth_req):
+    from journal import do_login
+    auth_req.params = {'username': 'bad', 'password': 'secret'}
+    assert not do_login(auth_req)
+
+
+def test_do_login_missing_params(auth_req):
+    from journal import do_login
+    for params in ({'username': 'admin'}, {'password': 'secret'}):
+        auth_req.params = params
+        with pytest.raises(ValueError):
+            do_login(auth_req)
+
+
+def login_helper(username, password, app):
+    """encapsulate app login for reuse in tests
+
+    Accept all status codes so that we can make assertions in tests
+    """
+    login_data = {'username': username, 'password': password}
+    return app.post('/login', params=login_data, status='*')
+
+
+def test_start_as_anonymous(app):
+    response = app.get('/', status=200)
+    actual = response.body
+    soup = BeautifulSoup(actual)
+    assert not soup.find(id='new-entry-btn')
+
+
+def test_login_success(app):
+    username, password = ('admin', 'secret')
+    redirect = login_helper(username, password, app)
+    assert redirect.status_code == 302
+    response = redirect.follow()
+    assert response.status_code == 200
+    actual = response.body
+    soup = BeautifulSoup(actual)
+    assert soup.find(id='new-entry-btn')
+
+
+def test_login_fails(app):
+    username, password = ('admin', 'wrong')
+    response = login_helper(username, password, app)
+    assert response.status_code == 200
+    actual = response.body
+    assert "Login Failed" in actual
+    soup = BeautifulSoup(actual)
+    assert not soup.find(id='new-entry-btn')
+
+
+def test_logout(app):
+    # re-use existing code to ensure we are logged in when we begin
+    test_login_success(app)
+    redirect = app.get('/logout', status="3*")
+    response = redirect.follow()
+    assert response.status_code == 200
+    actual = response.body
+    soup = BeautifulSoup(actual)
+    assert not soup.find(id='new-entry-btn')
+
+
+def test_new_entry_page_exists(app):
+    test_login_success(app)
+    response = app.get('/new_entry', status=200)
+    actual = response.body
+    soup = BeautifulSoup(actual)
+    assert soup.find(id='post-entry-btn')
+    assert soup.find('form')
+
+
+def test_no_new_entry_if_unauthn(app):
+    response = app.get('/new_entry', status=200)
+    actual = response.body
+    soup = BeautifulSoup(actual)
+    assert not soup.find('form')
```

Comment:
Good start on additional tests.  You need to push a bit harder to think of ways to hack your application.